transformer_engine/common/layer_norm/norms.h

```diff
@@ -0,0 +1,334 @@
+        dgamma_part(nullptr),
+        dx(nullptr),
+        dbeta(nullptr),
+        dgamma(nullptr) {}
+
+  // Input: gradient wrt. LN FWD output.
+  void* dz;
+
+  // Workspace for Wgrad pre-reduction.
+  void* dbeta_part;
+  void* dgamma_part;
+
+  // Output: Dgrad.
+  void* dx;
+  // Output: Wgrad.
+  void* dbeta;
+  void* dgamma;
+};
+
+enum NVTE_NORM_TYPE {
+  LN_FWD_TE,
+  LN_BWD_TE,
+  LN_FWD_CUDNN,
+  LN_BWD_CUDNN,
+  RMS_FWD_TE,
+  RMS_BWD_TE,
+  RMS_FWD_CUDNN,
+  RMS_BWD_CUDNN,
+};
+
+template <NVTE_NORM_TYPE NormEnum>
+constexpr bool IF_TE_NORMS() {
+  return (NormEnum == NVTE_NORM_TYPE::LN_FWD_TE || NormEnum == NVTE_NORM_TYPE::LN_BWD_TE ||
+          NormEnum == NVTE_NORM_TYPE::RMS_FWD_TE || NormEnum == NVTE_NORM_TYPE::RMS_BWD_TE);
+};
+
+template <NVTE_NORM_TYPE NormEnum>
+constexpr bool IF_TE_FWD_NORMS() {
+  return (NormEnum == NVTE_NORM_TYPE::LN_FWD_TE || NormEnum == NVTE_NORM_TYPE::RMS_FWD_TE);
+};
+
+template <NVTE_NORM_TYPE NormEnum>
+constexpr bool IF_TE_BWD_NORMS() {
+  return (NormEnum == NVTE_NORM_TYPE::LN_BWD_TE || NormEnum == NVTE_NORM_TYPE::RMS_BWD_TE);
+};
+
+using FwdFunction = std::function<void(LaunchParams<FwdParams>&, const bool)>;
+using BwdFunction = std::function<void(LaunchParams<BwdParams>&, const bool)>;
+using FunctionKey = uint64_t;
+using FwdTunedRegistry = std::unordered_map<FunctionKey, FwdFunction>;
+using BwdTunedRegistry = std::unordered_map<FunctionKey, BwdFunction>;
+using FwdGeneralRegistry = std::unordered_map<FunctionKey, std::map<uint64_t, FwdFunction>>;
+using BwdGeneralRegistry = std::unordered_map<FunctionKey, std::map<uint64_t, BwdFunction>>;
+
+template <NVTE_NORM_TYPE NormEnum, typename Enable = void>
+struct LauncherType;
+
+template <NVTE_NORM_TYPE NormEnum>
+struct LauncherType<NormEnum, typename std::enable_if<IF_TE_NORMS<NormEnum>()>::type> {
+  using ParamsType = std::conditional_t<IF_TE_FWD_NORMS<NormEnum>(), LaunchParams<FwdParams>,
+                                        LaunchParams<BwdParams>>;
+  using FunctionType = std::conditional_t<IF_TE_FWD_NORMS<NormEnum>(), FwdFunction, BwdFunction>;
+};
+
+extern FwdTunedRegistry LN_FWD_TUNED_FUNCS;
+extern BwdTunedRegistry LN_BWD_TUNED_FUNCS;
+extern FwdGeneralRegistry LN_FWD_GENERAL_FUNCS;
+extern BwdGeneralRegistry LN_BWD_GENERAL_FUNCS;
+
+extern FwdTunedRegistry RMS_FWD_TUNED_FUNCS;
+extern BwdTunedRegistry RMS_BWD_TUNED_FUNCS;
+extern FwdGeneralRegistry RMS_FWD_GENERAL_FUNCS;
+extern BwdGeneralRegistry RMS_BWD_GENERAL_FUNCS;
+
+template <NVTE_NORM_TYPE NormEnum, bool IF_TUNED, typename Enable = void>
+struct RegistryType {};
+
+template <NVTE_NORM_TYPE NormEnum, bool IF_TUNED>
+struct RegistryType<NormEnum, IF_TUNED, typename std::enable_if<IF_TE_NORMS<NormEnum>()>::type> {
+  using type = std::conditional_t<
+      IF_TUNED, std::conditional_t<IF_TE_FWD_NORMS<NormEnum>(), FwdTunedRegistry, BwdTunedRegistry>,
+      std::conditional_t<IF_TE_FWD_NORMS<NormEnum>(), FwdGeneralRegistry, BwdGeneralRegistry>>;
+};
```

Comment:
While templating on `IF_TUNED` makes sense for our current kernel implementation, I don't think it's a good idea going forward. The reason TE compilation is so slow (and is unusable on some systems) is because we statically compile so many tuned LayerNorm/RMSNorm kernels. If we choose to keep the TE tuned kernels, we should port them to NVRTC so we can compile them at runtime, similar to how we handle the transpose kernels. This would completely remove the need for this static registry for tuned kernels.